File: services/php/AuthMiddleware.php
<?php

namespace App\Middleware;

class AuthMiddleware
{
    private $tokenManager;
    private $publicPaths = ['/auth/login', '/auth/register', '/health'];

    public function __construct($tokenManager)
    {
        $this->tokenManager = $tokenManager;
    }

    public function handle($request, $next)
    {
        $path = $request['path'] ?? $_SERVER['REQUEST_URI'];

        if ($this->isPublicPath($path)) {
            return $next($request);
        }

        $token = $this->extractToken();

        if (!$token) {
            return $this->unauthorizedResponse('No token provided');
        }

        $result = $this->tokenManager->verify($token);

        if (!$result['valid']) {
            return $this->unauthorizedResponse('Invalid token');
        }

        $request['current_user'] = $result['payload'];
        return $next($request);
    }

    private function extractToken()
    {
        $authHeader = $_SERVER['HTTP_AUTHORIZATION'] ?? '';

        if (preg_match('/Bearer\s+(.+)/', $authHeader, $matches)) {
            return $matches[1];
        }

        return null;
    }

    private function isPublicPath($path)
    {
        foreach ($this->publicPaths as $publicPath) {
            if (strpos($path, $publicPath) === 0) {
                return true;
            }
        }
        return false;
    }

    private function unauthorizedResponse($message)
    {
        http_response_code(401);
        header('Content-Type: application/json');
        echo json_encode(['error' => $message]);
        exit;
    }
}

File: services/ruby/middleware/auth_middleware.rb
module Middleware
  class AuthMiddleware
    def initialize(app, token_manager)
      @app = app
      @token_manager = token_manager
    end

    def call(env)
      return @app.call(env) if public_path?(env['PATH_INFO'])

      token = extract_token(env)
      
      unless token
        return [401, { 'Content-Type' => 'application/json' }, 
                [{ error: 'No token provided' }.to_json]]
      end

      result = @token_manager.verify(token)
      
      if result.success?
        env['current_user'] = result.value
        @app.call(env)
      else
        [401, { 'Content-Type' => 'application/json' }, 
         [{ error: 'Invalid token' }.to_json]]
      end
    end

    private

    def extract_token(env)
      auth_header = env['HTTP_AUTHORIZATION']
      return nil unless auth_header
      
      match = auth_header.match(/^Bearer (.+)$/)
      match[1] if match
    end

    def public_path?(path)
      public_paths = ['/auth/login', '/auth/register', '/health']
      public_paths.any? { |p| path.start_with?(p) }
    end
  end
end
